File: source/ch/cyberduck/core/local/FinderSidebarService.java
package ch.cyberduck.core.local;

/*
 * Copyright (c) 2002-2015 David Kocher & Yves Langisch. All rights reserved.
 * https://cyberduck.io/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to feedback@cyberduck.io
 */

import ch.cyberduck.core.Local;
import ch.cyberduck.core.exception.LocalAccessDeniedException;
import ch.cyberduck.core.library.Native;

import org.apache.log4j.Logger;

import java.io.File;
import java.io.IOException;

/**
 * @version $Id:$
 */
public class FinderSidebarService implements SidebarService {
    private static final Logger log = Logger.getLogger(FinderSidebarService.class);

    static {
        Native.load("findersidebarservice");
    }

    private final String kLSSharedFileListFavoriteVolumes = "com.apple.LSSharedFileList.FavoriteVolumes";
    private final String kLSSharedFileListRecentServerItems = "com.apple.LSSharedFileList.RecentServers";
    private final String kLSSharedFileListFavoriteItems = "com.apple.LSSharedFileList.FavoriteItems";

    @Override
    public void add(final Local file) throws LocalAccessDeniedException {
        this.add(file, List.favorite);
    }

    public void add(final Local file, final List list) throws LocalAccessDeniedException {
        if(log.isDebugEnabled()) {
            log.debug(String.format("Add %s", file));
        }
        if(!this.addItem(file.getAbsolute(), this.forList(list))) {
            throw new LocalAccessDeniedException(String.format("Failure adding %s to %s", file, this.forList(list)));
        }
    }

    @Override
    public void remove(final Local file) throws LocalAccessDeniedException {
        this.remove(file, List.favorite);
    }

    public void remove(final Local file, final List list) throws LocalAccessDeniedException {
        if(log.isDebugEnabled()) {
            log.debug(String.format("Remove %s", file));
        }
        try {
            if(!this.removeItem(new File(file.getAbsolute()).getCanonicalPath(), this.forList(list))) {
                throw new LocalAccessDeniedException(String.format("Failure removing %s from %s", file, this.forList(list)));
            }
        }
        catch(IOException e) {
            throw new LocalAccessDeniedException(String.format("Failure removing %s from %s", file, this.forList(list)), e);
        }
    }

    private String forList(final List list) {
        switch(list) {
            case volume:
                return kLSSharedFileListFavoriteVolumes;
            case server:
                return kLSSharedFileListRecentServerItems;
        }
        return kLSSharedFileListFavoriteItems;
    }

    /**
     * @param file Path
     * @param list Shared item list name
     * @return False on failure
     */
    private native boolean addItem(final String file, final String list);

    /**
     * @param file Path
     * @param list Shared item list name
     * @return False on failure
     */
    private native boolean removeItem(final String file, final String list);
}

File: source/ch/cyberduck/core/local/FinderSidebarService.m
/*
 *  Copyright (c) 2005 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

#import <FinderSidebarService.h>
#import <CoreServices/CoreServices.h>
#import <Foundation/Foundation.h>
#import <JavaNativeFoundation/JNFString.h>

JNIEXPORT jboolean JNICALL Java_ch_cyberduck_core_local_FinderSidebarService_addItem(JNIEnv *env, jobject this, jstring file, jstring name) {
    LSSharedFileListRef list = LSSharedFileListCreate(kCFAllocatorDefault, (CFStringRef)JNFJavaToNSString(env, name), NULL);
    if (!list) {
		NSLog(@"Error getting shared file list reference");
        return NO;
    }
    NSURL* url = [NSURL fileURLWithPath:JNFJavaToNSString(env, file)];
    LSSharedFileListItemRef item = LSSharedFileListInsertItemURL(list,
                                                                 kLSSharedFileListItemLast,
                                                                 NULL, NULL,
                                                                 (CFURLRef)url,
                                                                 NULL, NULL);
    CFRelease(list);
    if(!item) {
        NSLog(@"Error getting shared file list item reference");
        return NO;
    }
    CFRelease(item);
    return YES;
}

JNIEXPORT jboolean JNICALL Java_ch_cyberduck_core_local_FinderSidebarService_removeItem(JNIEnv *env, jobject this, jstring file, jstring name) {
    LSSharedFileListRef list = LSSharedFileListCreate(kCFAllocatorDefault, (CFStringRef)JNFJavaToNSString(env, name), NULL);
    if (!list) {
        NSLog(@"Error getting shared file list reference");
        return NO;
    }
    UInt32 seed;
    CFArrayRef items = LSSharedFileListCopySnapshot(list, &seed);
    if (!items) {
        NSLog(@"Error getting shared file list items snapshot copy reference");
        return NO;
    }
    OSStatus err;
    for (CFIndex i = 0; i < CFArrayGetCount(items); i++) {
        LSSharedFileListItemRef item = (LSSharedFileListItemRef)CFArrayGetValueAtIndex(items, i);
        NSURL *url;
        if (noErr == (err = LSSharedFileListItemResolve(item, kLSSharedFileListNoUserInteraction, (CFURLRef*)&url, NULL))) {
            NSString *itemPath = [url path];
            [url release];
            if ([itemPath isEqualToString:JNFJavaToNSString(env, file)]) {
                if(noErr == (err = LSSharedFileListItemRemove(list, item))) {
                    break;
                }
                else {
                    NSLog(@"Error removing shared file list item. %s", GetMacOSStatusErrorString(err));
                }
            }
        }
        else {
            NSLog(@"Error resolving shared file list item. %s", GetMacOSStatusErrorString(err));
        }
    }
    CFRelease(items);
    CFRelease(list);
	return err == noErr;
}